Fix IOS bug in media Element when using Popup

ne0rrmatrix opened the pull request:
- Bug fix for Media Element trying to add controls to wrong View Controller.

 ### Description of Change ###
- Adds checks to make sure video controls do not get added to wrong View Controller. This works if using Shell, Page, Navigation Page, Tabbed Page, Popup, CollectionView, and CarouselView


 ### Linked Issues ###
 - Fixes #1556 
 - Fixes #1421

 ### PR Checklist ###
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls <!-- Replace this link to the direct link to your Pull Request in the MicrosoftDocs/CommunityToolkit repo  -->


 ### Additional information ###

 Fixes CTD when using media element in a Popup. The incorrect Parent ViewController is being used when adding controls to parent view controller. A sample for popup has been added to samples for media element. This document has been edited since written to better explain the changes and what the fix does.

bijington commented:
@ne0rrmatrix thank you for submitting this. Sorry to ask this but is there an easy way to test this fix? I'm guessing there isn't anything in the sample that can be tested?

ne0rrmatrix commented:
@bijington you can load this PR in the quoted issues example repo.

bijington commented:
> @bijington you can load this PR in the quoted issues example repo. This fix requires targeting dotnet 8.0.x.

OK thanks. Do you think there is any value in us introducing a similar example in the sample application? I'm happy to assist getting it in as part of the testing.

brminnick commented:
> Do you think there is any value in us introducing a similar example in the sample application?

Absolutely! Always 💯 

Anytime we fix a bug, we should add an example of its reproduction to the Sample App. This helps us verify the bug fix and ensures that we won't have a regression when testing a future PR.

ne0rrmatrix commented:
@brminnick I will work on adding reproduction to the sample app this week? I assume adding with PR is a good idea? Or should it be seperate? I am thinking separate because if we end up reverting PR the example goes away?

brminnick commented:
Thanks! Yes, please add it to this PR

ne0rrmatrix commented:
@brminnick I rewrote some View Controller logic for IOS. This change allows the controls to actually work. I still have a bunch of testing to do so please wait to review. I added a Popup with mediaElement inside it to the Media element page as a button on the far right called Popup in the sample app.

ne0rrmatrix commented:
Have done some testing and fixed a bug with Navigation Page logic. I still have to test Flyout Page and Tabbed Page. Once I confirm they are working I will tag someone to review.

ne0rrmatrix commented:
Ok I have tested against shell, navigation Page, Content Page, and Tabbed Page. Does anyone have any other suggested tests?

ne0rrmatrix commented:
Fix for collection view controller when using Carousel View has been added. Anyone who looks at this for review please indicate whether the approach I use is appropriate. There is now at least one other PR out to fix same issue. I am not sure which approach is appropriate and what the maintainers feel is best to move forward. I am happy either way.

pictos commented on src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:
```diff
@@ -13,16 +15,15 @@ public class MauiMediaElement : UIView
-	/// <param name="parentViewController">The <see cref="UIViewController"/> that acts as the parent for <paramref name="playerViewController"/>.</param>
 	/// <exception cref="NullReferenceException">Thrown when <paramref name="playerViewController"/><c>.View</c> is <see langword="null"/>.</exception>
-	public MauiMediaElement(AVPlayerViewController playerViewController, UIViewController? parentViewController)
+	public MauiMediaElement(AVPlayerViewController playerViewController)
```
just to point out that this's a breaking change

ne0rrmatrix replied:
@pictos  can you give me an example?

ne0rrmatrix replied:
There might be a better solution: #1623 
@pictos if you can look at that PR and get back to me if that is a better solution that does not cause breaking changes? Someone else is working on a related problem that may make this PR moot.

pictos replied:
@ne0rrmatrix I'm not saying this is wrong or bad... Just point out that's a breaking change, so if this PR gets merged we will need to release a major version. What you can do is preserve the old ctor and mark it with the Obsolete attribute and not show it on Intellisense.

ne0rrmatrix replied:
Ok i will restore old ctor and mark obsolete attribute.

ne0rrmatrix replied:
@pictos  I reverted ctor and marked old ctor as ```[Obsolete]``` Hope I did it right. Please review and suggest any changes.

pictos commented on src/CommunityToolkit.Maui.MediaElement/Views/MauiMediaElement.macios.cs:
```diff
@@ -15,14 +17,15 @@ public class MauiMediaElement : UIView
 #if IOS16_0_OR_GREATER || MACCATALYST16_1_OR_GREATER
 		// On iOS 16+ and macOS 13+ the AVPlayerViewController has to be added to a parent ViewController, otherwise the transport controls won't be displayed.
-		var viewController = parentViewController ?? WindowStateManager.Default.GetCurrentUIViewController();
+		var viewController = WindowStateManager.Default.GetCurrentUIViewController();
```
Can you revert the changes on this ctor? This is still a breaking change.

ne0rrmatrix replied:
Yes I will do that now.

ne0rrmatrix replied:
@pictos Updated as requested
